ouroboros-consensus/docs/utxo-hd/report.tex

```diff
@@ -0,0 +1,741 @@
+it. Now only one more special case arises which was not needed prior to these
+changes: the disk anchor moving.
+
+If \texttt{queryDisk} fails to retrieve the values because the
+\texttt{DbChangelog} held by the mempool got out of sync with the one on the
+disk, we are forced to perform a sync with the ledger in order to get an up to
+date \texttt{DbChangelog}. As the anchor belongs to the immutable part of the
+ledger database, it can only move forward, either along the same chain that is
+being represented currently on the mempool, or on a different fork. In the first
+case, the solution is easy, just get the new \texttt{DbChangelog} from the
+ledger database. In the second case, the chain tip must have moved and therefore
+the background thread will make sure that we revalidate our whole mempool and
+will result also in retrieving an up to date \texttt{DbChangelog}.
+
+Note that in the first case, we don't need to revalidate the transactions in the
+mempool despite the changelog changing as it still represents the same logical
+value.
+
+\subsubsection{Syncing with the ledger}
+
+Syncing with the ledger is a fairly simple operation, only worth noting out
+because it holds the flushing lock on read mode to keep a consistent view while
+revalidating all the transactions.
+
+\begin{algorithm}
+  \caption{Syncing with the ledger database}
+  \label{alg:sync}
+  \begin{algorithmic}[1]
+    \Procedure{SyncWithLedger}{$DB :: \texttt{LedgerDB},M :: \texttt{Mempool}$}
+    \State $\texttt{acquireRead}(DB)$
+    \State $(L, \texttt{Diffs}) \gets \texttt{getBestLedger}_{DB}$
+    \State $M' \gets M [ \texttt{isLedgerState} \gets L, \texttt{isDbChangelog} \gets \texttt{Diffs}] $
+    \State $M'' \gets \texttt{fold}(\texttt{revalidateMempool}(DB), M', \texttt{isTxs}_{M})$
+    \State $\texttt{releaseRead(DB)}$
+    \State \Return $M''$
+    \EndProcedure
+ \end{algorithmic}
+\end{algorithm}
+
+\subsubsection{Getting snapshots for forging}
+
+When forging a block, the node kernel monitors the current slot number and on
+each change it runs the following (simplified) algorithm:
+
+\begin{algorithm}
+  \caption{Forging logic}
+  \begin{algorithmic}[1]
+    \Procedure{ForgeFor}{$s :: \texttt{Slot}, DB :: \texttt{LedgerDB}, M :: \texttt{Mempool}$}
+    \State $\texttt{acquireRead}(DB)$
+    \State $((H, L_{E}), \texttt{Diffs}) \gets \texttt{getBestLedger}_{DB}$
+    \State $\texttt{ledgerView} \gets \texttt{forecastFor}(\texttt{forecastAt}(L_{E}), s)$
+    \State $proof \gets \texttt{shouldForge}(\texttt{ledgerView}, \texttt{H})$
+    \If{$\neg proof$}
+      \State \Return
+      \EndIf
+    \State $L^{s}_{D(s)} \gets \texttt{TICK}(L_{E}, s)$
+    \State $\texttt{validTxs} \gets \texttt{getSnapshotFor}_{M}(L^{s}_{D(s)}, s, \texttt{Diffs})$
+    \State $\texttt{releaseRead(DB)}$
+    \State $B \gets \texttt{forge}(L^{s}_{E}, proof, \texttt{validTxs})$
+    \State $\texttt{addBlockSync(DB, B)}$
+    \EndProcedure
+ \end{algorithmic}
+\end{algorithm}
```

Comment:
AFAICT, the only thing that UTxO HD changed here is the locking -- so maybe leave out the rest?